### ISSUE TITLE
tensor._copy should not copy when src and dst point to the same memory

### ISSUE THREAD

**Issue by msbaines** [pytorch/pytorch#34525]
## 🐛 Bug

I noticed a few unexpected 1GB DeviceToDevice cudaMemcpyAsync while training my model. I root caused these copies to tensor._copy doing a cudaMemcpy even when src and dst are the same. According to the CUDA [documentation](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79) overlapping src and dst is not supported by cudaMemcpyAsync. So this should be considered undefined-behavior.

## To Reproduce

```
        x = torch.tensor([1., 2., 3.]).to(device="cuda:0")
        y = x.data.float()
        print(x.data_ptr())
        print(y.data_ptr())
        y[0] = 2.0
        with torch.autograd.profiler.record_function("copy-bug"):
            x.data.copy_(y)
        print(x)
```

<img width="749" alt="Screen Shot 2020-03-09 at 8 25 54 PM" src="https://user-images.githubusercontent.com/35972327/76276855-41d50880-6244-11ea-9bbf-abca5bf580b6.png">

## Expected behavior

There should be no cudaMemcpyAsync. Instead, the copy_ should be a no-op when src and dst are the same memory.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.105

Versions of relevant libraries:
[pip] numpy==1.17.4
[pip] torch==1.4.0
[pip] torchvision==0.5.0
[conda] blas                      1.0                         mkl
[conda] magma-cuda101             2.5.1                         1    pytorch
[conda] mkl                       2019.4                      243
[conda] mkl-include               2019.4                      243
[conda] mkl-service               2.3.0            py36he904b0f_0
[conda] mkl_fft                   1.0.15           py36ha843d7b_0
[conda] mkl_random                1.1.0            py36hd6b4f25_0
[conda] pytorch                   1.4.0           py3.6_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.5.0                py36_cu101    pytorch

## Additional context

cc @ngimel

**Comment by albanD** [comment]
The problem is that detecting if two Tensors overlap or not is not a simple problem. And it can be quite expensive to check it every time we do a copy.
I think in general the user should not call `.copy_()` if the two Tensors are the same no?
And especially, wrt the issue you linked, the conversion to a different type should be a no-op if the type is already correct as this is easy to check and unambiguous.

**Comment by msbaines** [comment]
> The problem is that detecting if two Tensors overlap or not is not a simple problem. And it can be quite expensive to check it every time we do a copy.

There seems to already be a check here for sameness:

https://github.com/pytorch/pytorch/blob/f5ee46f1cfea4bbdd9fbe00a1f50b4385c5e4faa/aten/src/ATen/native/Copy.cpp#L101

Perhaps is_same() could be extended. In my repro case, it's the same memory (i.e. totally overlapping).

We ran into this issue while adding support for FP16 in our optimizer. The optimizer casts the parameters to float, updates, and the copies back. For an FP16 tensor, float() would create a copy but for FP32 it does not make a copy unless you explicitly ask for one. It was only while profiling that we noticed the copy:

https://github.com/pytorch/fairseq/blob/a091ae5f3b58564f4aa05b83d26155add486335a/fairseq/optim/adam.py#L198

I will try to avoid the copy via:

```
if dst.data_ptr() != src.data_ptr():
    dst.data._copy(src)
```

**Comment by albanD** [comment]
So you mean that calling `.float()` on a Tensor that is already of type float still makes a copy??

**Comment by msbaines** [comment]
> So you mean that calling `.float()` on a Tensor that is already of type float still makes a copy??

No, `.float()` does not make a copy except when the tensor is FP16. The copy happens later when call `._copy()`. For the FP32 case, the `._copy()` should be a no-op. Our code is similar to the repro steps I included.

**Comment by ngimel** [comment]
For UB, when source and destination are the same, in practice results are correct, for partial overlap it's indeed UB. So functionally we are good here, but you are right, cudaMemcpy should not be invoked in this case.

**Comment by albanD** [comment]
You mean `.copy_()` right?

Note that for partial overlap, the result will be UB for cpu as well as we don't have a strict read/write ordering.

Also you should not use .data but you should wrap your code in `with torch.no_grad()` if you don't want gradients.
Once you do that, you can guard the `.copy_()` with `if x is not y:`.

**Comment by ngimel** [comment]
Use of `.data` is discouraged. As @albanD suggests, if you don't want some operations tracked by autograd, you should wrap them in `no_grad()` context manager. Once you do that, `x.copy_(y)` passes `is_same()` check and no memcopy is invoked.

**Comment by albanD** [comment]
Good point by @ngimel this check is already done in the `.copy_()` implementation [here](https://github.com/pytorch/pytorch/blob/cb689a5d68a9c4c203f0d3dffbb1ef2cb7d2a506/aten/src/ATen/native/Copy.cpp#L101). So if you don't use `.data` (which you should not), this will already happen properly!

As mentioned above, for the more general case where you have a different Tensor that is potentially an arbitrary view of the one you copy into, then you will have UB.

**Comment by msbaines** [comment]
> Use of `.data` is discouraged. As @albanD suggests, if you don't want some operations tracked by autograd, you should wrap them in `no_grad()` context manager. Once you do that, `x.copy_(y)` passes `is_same()` check and no memcopy is invoked.

This is great advice! I will try `no_grad()`.
